### PR TITLE
[RAPTOR-6106] Refactor prediction validation and marshalling

### PR DESCRIPTION
**Pull request by 4dahalibut**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
I moved predictions validation and marshalling from being different in different artifact/language predictors to being centralized in the base language predictor, and then I tested it better

## Rationale
While trying to straighten out class label marshalling, I decided that all the artifact predictors share logic that would better be done in one place. Now we pass raw labels downstream and do all marshalling in one place

### CODE REVIEW

File: tests/drum/test_data_marshalling.py [hunk 1]
```diff
@@ -27,6 +32,70 @@ def test__order_by_float():
...
+    ):
+        marshal_predictions(
+            request_labels=labels, predictions=preds, target_type=TargetType.MULTICLASS
+        )
+
+
+def test_marshal_predictions_binary_reshape_happy():
+    preds = np.array([1, 1, 1])
+    labels = [1, 2]
+    marshal_predictions(request_labels=labels, predictions=preds, target_type=TargetType.BINARY)
```
Comment (meltzg) [inline tests/drum/test_data_marshalling.py:62]: please test that the resulting predictions are correct

File: tests/drum/test_data_marshalling.py [hunk 2]
```diff
@@ -27,6 +32,70 @@ def test__order_by_float():
...
+
+
+def test_marshal_predictions_reshape_regression_happy():
+    preds = np.array([1, 1, 1])
+    labels = [1]
+    marshal_predictions(request_labels=labels, predictions=preds, target_type=TargetType.REGRESSION)
```
Comment (meltzg) [inline tests/drum/test_data_marshalling.py:68]: please test that the resulting predictions are correct

File: tests/drum/test_data_marshalling.py (hunk 1 above)
Comment (meltzg) [reply tests/drum/test_data_marshalling.py:62]: please add test for binary with both values supplied in each row

File: tests/drum/test_data_marshalling.py [hunk 3]
```diff
@@ -27,6 +32,70 @@ def test__order_by_float():
...
+    expected = pd.DataFrame(data=preds, columns=labels)
+    pd.testing.assert_frame_equal(
+        marshal_predictions(
+            request_labels=labels, predictions=preds, target_type=TargetType.MULTICLASS
+        ),
+        expected,
+    )
+
+
+def test_marshal_predictions_multiclass_wrong_label_length():
```
Comment (meltzg) [inline tests/drum/test_data_marshalling.py:47]: please add a test for num columns > num labels

File: tests/drum/test_data_marshalling.py [hunk 4]
```diff
@@ -27,6 +32,70 @@ def test__order_by_float():
...
+def test_marshal_predictions_bad_shape_regression():
+    labels = [1]
+    preds = np.zeros((1, 2))
+    with pytest.raises(DrumCommonException, match="must contain only 1 column"):
+        marshal_predictions(
+            request_labels=labels, predictions=preds, target_type=TargetType.REGRESSION
+        )
+
+
+def test_marshal_predictions_dont_add_to_one():
```
Comment (meltzg) [inline tests/drum/test_data_marshalling.py:95]: please add a test for invalid other ways of adding to 1 (ex. 2, -1)

**Review by meltzg — CHANGES_REQUESTED** [review]
needs more test cases

File: tests/drum/test_data_marshalling.py (hunk 4 above)
Comment (4dahalibut) [reply tests/drum/test_data_marshalling.py:95]: how come this is bad?

Comment (meltzg) [reply tests/drum/test_data_marshalling.py:95]: because 2, -1 isn't a valid probability distribution but it still sums to 1

Comment (4dahalibut) [reply tests/drum/test_data_marshalling.py:95]: what is invalid about it? the fact there is a negative? I can assert no negatives if you'd like

Comment (meltzg) [reply tests/drum/test_data_marshalling.py:95]: Yes, as insinuated in my previous comment, -1 is not a valid probability. Probabilities should be in [0.0, 1.0]

File: custom_model_runner/datarobot_drum/drum/data_marshalling.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        pass
+
+    # Okay lets just do a str.lower
+    assert isinstance(label, str)
+    return label.lower()
+
+
+def _classification_marshal_preds(predictions, request_labels, model_labels):
+    request_labels = _marshal_labels(request_labels, model_labels)
+    predictions = _duplicate_1d_probabilities(predictions, request_labels)
```
Comment (meltzg) [inline custom_model_runner/datarobot_drum/drum/data_marshalling.py:94]: nit: _infer_negative_class_probabilities or something like that is probably a better name, but not enough to warrant rerunning tests

**Review by meltzg — APPROVED** [review]
minor naming nit, but otherwise, great job. Glad to see code made common

File: custom_model_runner/datarobot_drum/drum/data_marshalling.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        raise DrumCommonException(
+            f"Regression and anomaly predictions must contain only 1 column. Your predictions have {predictions.shape[1]} columns"
+        )
+
+
+def _validate_probabilities_make_sense(predictions):
+    if (predictions < 0).any():
+        raise DrumCommonException("Your predictions have negative values")
+    if (predictions > 1).any():
+        raise DrumCommonException("Your predictions have values greater than 1")
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/data_marshalling.py:130]: nit: Your prediction probabilities

File: custom_model_runner/datarobot_drum/drum/data_marshalling.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    if predictions.shape[1] != len(labels_to_use):
+        raise DrumCommonException(
+            "Your predictions must return the "
+            "probability distribution for the correct number of class labels. "
+            "Expected {} columns, but received {}".format(len(labels_to_use), predictions.shape[1])
+        )
+
+
+def _duplicate_1d_probabilities(predictions, labels):
+    if predictions.shape[1] == 1:
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/data_marshalling.py:150]: we should make sure target type is binary here

File: custom_model_runner/datarobot_drum/drum/data_marshalling.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+
+def _validate_predictions_are_one_dimensional(predictions):
+    if predictions.shape[1] != 1:
+        raise DrumCommonException(
+            f"Regression and anomaly predictions must contain only 1 column. Your predictions have {predictions.shape[1]} columns"
+        )
+
+
+def _validate_probabilities_make_sense(predictions):
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/drum/data_marshalling.py:126]: nit: _validate_probabilities_sum_to_one

**Review by jmbannon — APPROVED** [review]

File: custom_model_runner/datarobot_drum/drum/data_marshalling.py (hunk 7 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/data_marshalling.py:150]: we actually can duplicate for the multiclass case too where there are only 2 labels.

**Pull request merged**